pulsar-connect/twitter/src/main/java/org/apache/pulsar/connect/twitter/TwitterFireHose.java

```diff
@@ -0,0 +1,172 @@
+    private Function<String, CompletableFuture<Void>> consumeFunction;
+
+    @Override
+    public void open(Map<String, String> config) {
+        verifyExists(config, CONSUMER_KEY);
+        verifyExists(config, CONSUMER_SECRET);
+        verifyExists(config, TOKEN);
+        verifyExists(config, TOKEN_SECRET);
+        waitObject = new Object();
+        startThread(config);
+    }
+
+    @Override
+    public void setConsumer(Function<String, CompletableFuture<Void>> consumeFunction) {
+        this.consumeFunction = consumeFunction;
+    }
+
+    @Override
+    public void close() throws Exception {
+        stopThread();
+    }
+
+
+    private void verifyExists(final Map<String, String> config, final String key) {
+        if (!config.containsKey(key)) {
+            throw new IllegalArgumentException("Required property '" + key + "' not set.");
+        }
+    }
+    // ------ Custom endpoints
+
+    /**
+     * Implementing this interface allows users of this source to set a custom endpoint.
+     */
+    public interface EndpointInitializer {
+        StreamingEndpoint createEndpoint();
+    }
+
+    /**
+     * Required for Twitter Client
+     */
+    private static class SampleStatusesEndpoint implements EndpointInitializer, Serializable {
+        @Override
+        public StreamingEndpoint createEndpoint() {
+            // this default endpoint initializer returns the sample endpoint: Returning a sample from the firehose (all tweets)
+            StatusesSampleEndpoint endpoint = new StatusesSampleEndpoint();
+            endpoint.stallWarnings(false);
+            endpoint.delimited(false);
+            return endpoint;
+        }
+    }
+
+    private void startThread(Map<String, String> config) {
+        Authentication auth = new OAuth1(config.get(CONSUMER_KEY),
+                config.get(CONSUMER_SECRET),
+                config.get(TOKEN),
+                config.get(TOKEN_SECRET));
+
+        BasicClient client = new ClientBuilder()
+                .name(config.getOrDefault(CLIENT_NAME, "openconnector-twitter-source"))
+                .hosts(config.getOrDefault(CLIENT_HOSTS, Constants.STREAM_HOST))
+                .endpoint(new SampleStatusesEndpoint().createEndpoint())
+                .authentication(auth)
+                .processor(new HosebirdMessageProcessor() {
+                    public DelimitedStreamReader reader;
+
+                    @Override
+                    public void setup(InputStream input) {
+                        reader = new DelimitedStreamReader(input, Constants.DEFAULT_CHARSET,
+                            Integer.parseInt(config.getOrDefault(CLIENT_BUFFER_SIZE, "50000")));
+                    }
+
+                    @Override
+                    public boolean process() throws IOException, InterruptedException {
+                        String line = reader.readLine();
+                        try {
+                            consumeFunction.apply(line);
```

Comment:
If `apply()` is returning a future (tied to publishing the message on a topic), then we should probably wait